File: jdacommandinterface-core/src/main/java/io/github/romangraef/jdacommandinterface/core/util/ArgumentUtil.java
package io.github.romangraef.jdacommandinterface.core.util;

import io.github.romangraef.jdacommandinterface.core.Context;
import io.github.romangraef.jdacommandinterface.core.ConversionException;
import io.github.romangraef.jdacommandinterface.core.NoConverterFoundException;
import io.github.romangraef.jdacommandinterface.core.NotEnoughArgumentsException;
import io.github.romangraef.jdacommandinterface.core.converters.Converters;

import java.lang.reflect.Array;
import java.lang.reflect.Method;
import java.lang.reflect.Parameter;

public class ArgumentUtil {
    /**
     * For internal use. Converts a string list into a list of objects according to a method definition.
     *
     * if varargs is set to true the last element of the result will be an array of those converted varargs
     *
     * @param context   the context of the command as passed to the {@link Converters}
     * @param args      the arguments to convert
     * @param argCount  the argument count of the method definition excluding {@link Context} and varargs
     * @param isVarArgs whether the method has varargs.
     * @param method    the method definition to use
     *
     * @return the arguments converted to objects.
     * @throws NoConverterFoundException if the method asks for an object without a registered converter
     * @throws ConversionException this is reraised from the converters and indicated some kind of user error
     * @throws NotEnoughArgumentsException if the provided arguments aren't enough to fulfill the needs of the method.
     */
    public static Object[] getArguments(Context context, String[] args, int argCount, boolean isVarArgs, Method method) throws NoConverterFoundException, ConversionException, NotEnoughArgumentsException {
        Object[] finalArgs = new Object[argCount + 1 + (isVarArgs ? 1 : 0)];
        finalArgs[0] = context;
        Object[] normalArgs = getConvertedNormalArgs(args, argCount, method, context);
        System.arraycopy(normalArgs, 0, finalArgs, 1, normalArgs.length);
        if (isVarArgs) {
            Parameter p = null;
            for (Parameter pa : method.getParameters()) {
                if (pa.isVarArgs()) {
                    p = pa;
                    break;
                }
            }
            Class<?> clazz = p.getType().getComponentType();
            Object[] varArgs = (Object[]) Array.newInstance(clazz, args.length - argCount);

            for (int i = argCount, j = 0; j < args.length - argCount; i++, j++) {
                varArgs[j] = Converters.convert(clazz, context, args[i]);
            }
            finalArgs[argCount + 1] = varArgs;
        }
        return finalArgs;
    }

    private static Object[] getConvertedNormalArgs(String[] args, int argCount, Method method, Context context) throws NoConverterFoundException, ConversionException, NotEnoughArgumentsException {
        Object[] converted = new Object[argCount];
        Parameter[] parameters = method.getParameters();
        for (int i = 1; i < parameters.length; i++) {
            Parameter p = parameters[i];
            if (p.isVarArgs()) continue;
            try {
                converted[i - 1] = Converters.convert(p.getType(), context, args[i - 1]);
            } catch (ArrayIndexOutOfBoundsException e) {
                throw new NotEnoughArgumentsException(e);
            }
        }
        return converted;
    }
}

File: jdacommandinterface-core/src/main/java/io/github/romangraef/jdacommandinterface/core/util/ArgumentUtil.kt
package io.github.romangraef.jdacommandinterface.core.util

import io.github.romangraef.jdacommandinterface.core.Context
import io.github.romangraef.jdacommandinterface.core.ConversionException
import io.github.romangraef.jdacommandinterface.core.NoConverterFoundException
import io.github.romangraef.jdacommandinterface.core.NotEnoughArgumentsException
import io.github.romangraef.jdacommandinterface.core.converters.Converters
import java.lang.reflect.Parameter

object ArgumentUtil {
    /**
     * For internal use. Converts a string list into a list of objects according to a method definition.
     *
     * @param context    the context of the command as passed to the [Converters]
     * @param args       the arguments to convert
     * @param parameters the parameters of the method
     * @return the arguments converted to objects.
     * @throws NoConverterFoundException   if the method asks for an object without a registered converter
     * @throws ConversionException         this is reraised from the converters and indicated some kind of user error
     * @throws NotEnoughArgumentsException if the provided arguments aren't enough to fulfill the needs of the method.
     */
    @JvmStatic
    @Throws(NoConverterFoundException::class, ConversionException::class, NotEnoughArgumentsException::class)
    fun getArguments(context: Context, args: Array<String>, parameters: Array<Parameter>): Array<Any?> {
        val finalArgs = arrayOfNulls<Any>(parameters.size)
        // First argument is the context
        finalArgs[0] = context

        // Convert rest arguments
        val normalArgs = getConvertedNormalArgs(args, parameters.copyOfRange(1, parameters.size), context)

        // Copy rest arguments into the final args array
        System.arraycopy(normalArgs, 0, finalArgs, 1, normalArgs.size)

        // Get Vararg
        val varArgParam = parameters
                .find { obj: Parameter -> obj.isVarArgs }

        if (varArgParam != null) {

            // Get Vararg type
            val clazz: Class<*> = varArgParam.type.componentType!!
            // Get String var args
            val varArgStrings: Array<String> = args.sliceArray(parameters.size - 1..args.size)

            // Convert those var args using Converters
            val varArgs: Array<Any?> = varArgStrings.map {
                Converters.convert(clazz, context, it)
            }.toTypedArray()

            // Save the vararg to the array
            finalArgs[parameters.size - 1] = varArgs
        }
        return finalArgs
    }

    @Throws(NoConverterFoundException::class, ConversionException::class, NotEnoughArgumentsException::class)
    private fun getConvertedNormalArgs(args: Array<String>,
                                       parameters: Array<Parameter>,
                                       context: Context): Array<Any?> = try {
        parameters
                .drop(1) // No Context
                .filter { !it.isVarArgs }
                .mapIndexed { index, p -> Converters.convert(p.type, context, args[index]) }
                .toTypedArray()
    } catch (e: ArrayIndexOutOfBoundsException) {
        throw NotEnoughArgumentsException(e)
    }

}